### PR TITLE
HPCC-25843 Validate data planes in WsWorkunits, etc

### PR DESCRIPTION
**Pull request by wangkx**
1. validate data planes in WsWorkunits.
2. validate data planes in WsPackageMap.
3. other 2 fixes.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: esp/smc/SMCLib/TpWrapper.cpp [hunk 1]
```diff
@@ -2516,3 +2516,48 @@ StringBuffer & getRoxieDefaultPlane(StringBuffer & plane, const char * roxieName
...
+    if (validDataPlaneNames.find(dataPlaneName) == validDataPlaneNames.end())
+    {
+        validDataPlaneNames.insert(dataPlaneName);
+        PROGLOG("adding valid data plane name: %s", dataPlaneName);
+    }
+}
+
+static void refreshValidDataPlaneNames()
+{
+    validDataPlaneNames.clear();
```
Comment (afishbeck) [inline esp/smc/SMCLib/TpWrapper.cpp:2537]: is this thread safe?

**Review by afishbeck — COMMENTED** [review]
Question about thread safety.

File: esp/smc/SMCLib/TpWrapper.cpp (hunk 1 above)
Comment (wangkx) [reply esp/smc/SMCLib/TpWrapper.cpp:2537]: Will change

**Comment by wangkx** [comment]
@afishbeck the code for thread safe is added. Please take a look and let me know when I should squash the commits and rebase to 8.2.0.

**Review by afishbeck — DISMISSED** [review]
@wangkx looks ok.  Please squash.  @richardkchapman should have input on whether to retarget.

**Comment by richardkchapman** [comment]
Yes, should target 8.2.0

File: esp/smc/SMCLib/TpContainer.cpp [hunk 2]
```diff
@@ -743,11 +743,36 @@ static void refreshValidTargets()
...
+    {
+        validDataPlaneNames.insert(dataPlaneName);
+        PROGLOG("adding valid data plane name: %s", dataPlaneName);
+    }
+}
+
+static void refreshValidDataPlaneNames()
+{
+    validDataPlaneNames.clear();
+    Owned<IPropertyTreeIterator> queues = getComponentConfigSP()->getElements("queues");
```
Comment (ghalliday) [inline esp/smc/SMCLib/TpContainer.cpp:761]: I'm not sure what this code is here for.  I think all the planes should be covered by the second iterator.  Did you find an example that wasn't?

File: esp/services/ws_packageprocess/ws_packageprocessService.cpp [hunk 3]
```diff
@@ -330,7 +330,7 @@ class PackageMapUpdater
 #else
```
Comment (ghalliday) [inline esp/services/ws_packageprocess/ws_packageprocessService.cpp:330]: Did you mean to remove the #ifdef?

File: esp/smc/SMCLib/TpContainer.cpp [hunk 4]
```diff
@@ -743,11 +743,36 @@ static void refreshValidTargets()
...
 
+static void addUniqueDataPlaneName(const char * dataPlaneName)
+{
+    if (isEmptyString(dataPlaneName))
+        return;
+
+    if (validDataPlaneNames.find(dataPlaneName) == validDataPlaneNames.end())
+    {
+        validDataPlaneNames.insert(dataPlaneName);
+        PROGLOG("adding valid data plane name: %s", dataPlaneName);
```
Comment (ghalliday) [inline esp/smc/SMCLib/TpContainer.cpp:754]: This could be a bit too verbose.

File: esp/smc/SMCLib/TpContainer.cpp [hunk 5]
```diff
@@ -756,12 +781,22 @@ extern TPWRAPPER_API void validateTargetName(const char* target)
...
         throw makeStringExceptionV(ECLWATCH_INVALID_CLUSTER_NAME, "Invalid target name: %s", target);
 }
 
+extern TPWRAPPER_API bool validateDataPlaneName(const char * remoteDali, const char * name)
+{
+    if (isEmptyString(name))
+        throw makeStringException(ECLWATCH_INVALID_CLUSTER_NAME, "Empty data plane name.");
+
+    CriticalBlock block(configUpdateSect);
+    configUpdateHook.installOnce(configUpdate, true);
```
Comment (ghalliday) [inline esp/smc/SMCLib/TpContainer.cpp:796]: better if the installOnce is outside the critical block - same for the call above on line 785

File: esp/smc/SMCLib/TpContainer.cpp [hunk 6]
```diff
@@ -743,11 +743,36 @@ static void refreshValidTargets()
     }
 }
 
+static void addUniqueDataPlaneName(const char * dataPlaneName)
+{
+    if (isEmptyString(dataPlaneName))
```
Comment (ghalliday) [inline esp/smc/SMCLib/TpContainer.cpp:748]: I think if you are only iterating the storage.planes this will always be non null and unique.

**Review by ghalliday — COMMENTED** [review]
@wangkx A few fairly minor comments.

File: esp/smc/SMCLib/TpContainer.cpp (hunk 2 above)
Comment (wangkx) [reply esp/smc/SMCLib/TpContainer.cpp:761]: I did not. But, I saw similar code from the existing getRoxieDefaultPlane() in TpWrapper.cpp. Please let me know if I should remove this code.

File: esp/services/ws_packageprocess/ws_packageprocessService.cpp (hunk 3 above)
Comment (wangkx) [reply esp/services/ws_packageprocess/ws_packageprocessService.cpp:330]: Yes, it should be removed. The problem happened when I rebased the code to 8.2.0. I found the same problem in another place.

File: esp/smc/SMCLib/TpContainer.cpp (hunk 4 above)
Comment (wangkx) [reply esp/smc/SMCLib/TpContainer.cpp:754]: How about changing to: 'adding data plane: %s'?

File: esp/smc/SMCLib/TpContainer.cpp (hunk 5 above)
Comment (wangkx) [reply esp/smc/SMCLib/TpContainer.cpp:796]: Ok, I see that installOnce() is thread safe.

File: esp/smc/SMCLib/TpContainer.cpp (hunk 6 above)
Comment (wangkx) [reply esp/smc/SMCLib/TpContainer.cpp:748]: @ghalliday please see my comment https://github.com/hpcc-systems/HPCC-Platform/pull/14995#discussion_r648315662.

File: esp/smc/SMCLib/TpContainer.cpp (hunk 2 above)
Comment (ghalliday) [reply esp/smc/SMCLib/TpContainer.cpp:761]: That code is doing something different - it is finding out what the default data plane associated with a roxie is.
First it checks if the roxie has a specific data plane associated with it.  If not it picks the first data plane.  I suspect with recent changes that should be implemented slightly differently.

File: esp/smc/SMCLib/TpContainer.cpp (hunk 4 above)
Comment (ghalliday) [reply esp/smc/SMCLib/TpContainer.cpp]: It was more the number of lines than the text.  @afishbeck do you mind?

File: esp/smc/SMCLib/TpContainer.cpp (hunk 2 above)
Comment (wangkx) [reply esp/smc/SMCLib/TpContainer.cpp:761]: I will remove the code.

**Review by ghalliday — APPROVED** [review]
@wangkx please can you squash

File: esp/smc/SMCLib/TpContainer.cpp (hunk 4 above)
Comment (afishbeck) [reply esp/smc/SMCLib/TpContainer.cpp]: I'd rather not have the excessive logging.  If it ends up merged this way could streamline later.

**Comment by wangkx** [comment]
@ghalliday the log line is removed and the commits are squashed.

**Pull request merged**